lib/src/layer/polyline_layer.dart

```diff
@@ -39,66 +69,191 @@ class Polyline {
     this.strokeCap = StrokeCap.round,
     this.strokeJoin = StrokeJoin.round,
     this.useStrokeWidthInMeter = false,
+    this.hitKey,
   });
 
-  /// Used to batch draw calls to the canvas.
-  int get renderHashCode => Object.hash(
-      strokeWidth,
-      color,
-      borderStrokeWidth,
-      borderColor,
-      gradientColors,
-      colorsStop,
-      isDotted,
-      strokeCap,
-      strokeJoin,
-      useStrokeWidthInMeter);
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      (other is Polyline<TapKeyType> &&
+          listEquals(points, other.points) &&
+          strokeWidth == other.strokeWidth &&
+          color == other.color &&
+          borderStrokeWidth == other.borderStrokeWidth &&
+          borderColor == other.borderColor &&
+          listEquals(gradientColors, other.gradientColors) &&
+          listEquals(colorsStop, other.colorsStop) &&
+          isDotted == other.isDotted &&
+          strokeCap == other.strokeCap &&
+          strokeJoin == other.strokeJoin &&
+          useStrokeWidthInMeter == other.useStrokeWidthInMeter &&
+          hitKey == other.hitKey);
+
+  List<Object?> get _baseHashableValues => [
+        strokeWidth,
+        color,
+        borderStrokeWidth,
+        borderColor,
+        gradientColors,
+        colorsStop,
+        isDotted,
+        strokeCap,
+        strokeJoin,
+        useStrokeWidthInMeter,
+        hitKey,
+      ];
+
+  /// Used to batch draw calls to the canvas
+  int get renderHashCode => Object.hashAll(_baseHashableValues);
+
+  @override
```

Comment:
Why are we doing both: giving the polyline a key identity and a hash identity? Shouldn't either do it?

:nit: this should be slightly more expensive since you're allocating a few more lists. If you restore the original renderHashCode you should be able to do Object.hash(renderHashCode, _, __, ...), i.e. it should be fine to hash the hash at the end of the day that's what they do internally as well and we don't have any cryptographic requirements.

Reply:
Yeah, I was trying to avoid hashing twice, but I guess it's OK. Resolved.

Reply:
Did you respond to my first or my second paragraph? I'd guess the latter.

My bigger question remains: why two different ways of identifying polylines?

(as for the second paragraph and your response, you're already hashing hashes anyway. To illustrate my point, a simple hash function would look something like:

```dart
int hash2(Object a, Object b) => a.hashCode ^ b.hashCode << 13;
```
)

Reply:
Render hash code is used for batching and layering operations, and doesn't include the list of coordinates, because that wouldn't be helpful.

The normal hash code should include the points, because the equality method does, and they should match.

Reply:
Let me rephrase, why do we need hashCode and hitKey?

Reply:
I'm not sure I understand, they are intended for entirely different purposes.

`hitKey` is just some freeform metadata for the user to assign if they want, helpful to track a line back out when handling.
`hashCode` is for the equality stuff and all the other normal uses of `hashCode`, which is not directly influenced by `hitKey`.

`hashCode` cannot just be `hitKey`, because multiple lines could have the same key, which could cause issues for the `hashCode`, if two objects are not equal but share the same hash.

(And stepping one step into the future, we've discussed removing `hitKey` and the generic typing, and just forcing users to extend `Polyline`, but a) that makes for an unnecessarily lengthy constructor and b) a type cast would be required somewhere.)

Reply:
Ah, I see where you might be getting at. Using the hashCode to identify the Polyline instead of the key at the other end. That is not ideal because the user has no direct control over the hashCode, the hashCode cannot be used to store information any more advanced than an integer, and the reason above. Think of hashCode as entirely internal as you would normally.

Reply:
99% of the time when folks implement hashCode it's to use the Object as a key in a map.

```dart
final data = <Polyline, Data>{ /* ...*/ };

void onTap(PolylineHit hit) {
  doSomething(data[hit.lines.first]);
}
```

```dart
void onTap(PolylineHit hit) {
  doSomething(hit.lines.first.hitKey);
}
```

Feel very similar to me. Or did you add the `hashCode` for another reason? Either way, above options are both possible now.

Reply:
Ooo, that's a nice way to attach data that I guess is possible now. Maybe that alone is enough to remove `hitKey`, and that'll clean up the generic typing as well. Good idea, I'll do that.

Reply:
Removed `tapKey` in favour of this method. I'll change the docs.